### PR TITLE
Add a documentation page describing the different parts of the Openverse stack

### LINKED ISSUE #3201: Add a documentation page describing the different parts of our stack
## Description

We have some information regarding the different aspects of our stack in the top level README.md: https://github.com/WordPress/openverse#openverse

It would be useful to have a documentation page on our docs site that goes further in-depth about the various pieces of our stack, with references to the related documentation and code for each entity (e.g. API, ingestion server, catalog, frontend).

### PR DESCRIPTION
**Pull request by zealshah29**
## Fixes

Fixes #3201 by @AetherUnbound

## Description

I have added a documentation page on the docs site that goes further in-depth about the various pieces of the stack, with references to the related documentation and code for each entity.

For each of the following sections of the stack listed, we have: 

- Link to the folder for the service on GitHub
- Link to the documentation for the service on docs.openverse.org
- Primary language(s) for the project
- List of all of the tools/technologies/frameworks the services use along with links to each of its documentation

For example for the API, we have (respectively):

- summary of the service
- [Code](https://github.com/WordPress/openverse/tree/main/api)
- [Documentation](https://docs.openverse.org/api/index.html)
- Language: Python
- Tools & Frameworks:
   - [Django REST framework](https://www.django-rest-framework.org/)

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: documentation/general/stack.md
Comment (dhruvkb) [inline documentation/general/stack.md:26]: Also include Django above this.

File: documentation/general/stack.md
Comment (dhruvkb) [inline documentation/general/stack.md:19]: Docker is basically a dev + deployment mechanism and not a part of the stack's technologies so I recommend dropping that.

File: documentation/general/stack.md
Comment (dhruvkb) [inline documentation/general/stack.md:32]: HTML and CSS are kind-of universal to frontend dev so we can skip them. Also TypeScript is more appropriate here.

```suggestion
   - Language: TypeScript / Node.js
```

File: documentation/general/stack.md
Comment (dhruvkb) [inline documentation/general/stack.md:37]: Node.js is neither a tool nor a framework. It's a JS runtime, so it feels incorrect to list here.

As for TailwindCSS, it is a very small, not-noteworthy slice of the number of technologies we use in the frontend.

File: documentation/general/stack.md
Comment (dhruvkb) [inline documentation/general/stack.md:50]: These blank lines should be removed as the lint job in CI will complain.

**Review by dhruvkb — CHANGES_REQUESTED** [review]
@zealshah29 this is a great start, thanks for the contributions. I do have some notes that I've elaborated on below.

1. You need to lint the file so that the CI check for linting can pass. You can do so using the `just lint` recipe. Linting will also run automatically from then on as the lint recipe will download pre-commit and set up the Git hooks for you.

2. The newly added `stack.md` file needs to be added to the table of contents, which present is in the file https://github.com/WordPress/openverse/blob/3e57bfaff03d44b97e73ef5794ef0eda40b33749/documentation/general/index.md.

3. Automation is not technically a part of the stack as it's more concerned with general repository management. I would drop it but I'll defer to other maintainers for their views.

4. The one-line descriptions for the stack layers can be found in the main `README.md` file, it'll be good to re-use those here.

5. There are other components in our stack that are more general 
purpose and shared between these layers. These include

   - cache (Redis)
   - upstream and API database (PostgreSQL)
   - anaytics (third-party, Plausible)

   IMO these can, and should, be included in their own separate subsection.

File: documentation/general/stack.md
Comment (zealshah29) [reply documentation/general/stack.md:26]: Shall I add it to line number 25?

File: documentation/general/stack.md
Comment (zealshah29) [reply documentation/general/stack.md:19]: Okay

File: documentation/general/stack.md
Comment (zealshah29) [reply documentation/general/stack.md:32]: Okay

File: documentation/general/stack.md
Comment (zealshah29) [reply documentation/general/stack.md:37]: Okay

File: documentation/general/stack.md
Comment (zealshah29) [reply documentation/general/stack.md:50]: Okay

File: documentation/general/stack.md
Comment (dhruvkb) [reply documentation/general/stack.md:26]: Yes, that would be great, thanks!

**Comment by dhruvkb** [comment]
> Hi, can you share the documentation for step-by-step guidance on how to lint the file?

I search our docs and it turns out unfortunately, we do not have docs for that, which is a big miss on our end (and a potential opportunity for you to submit a PR for, once you've set it up for yourself).

Linting locally requires a bunch of tooling set up like [`just`](https://docs.openverse.org/general/general_setup.html#just), [Docker](https://docs.openverse.org/general/general_setup.html#docker) and [Node.js](https://docs.openverse.org/general/general_setup.html#node-js). Would you be open to installing them? If so, please do so and then run `just lint` from the project root directory.

If you would like to not set them up, let us know and we'll lint the file for you.

> Shall I create a separate subsection for this?

Yes, a separate section with a name like "External services" would be good to group these services. All of our external services are in the https://github.com/WordPress/openverse/tree/b4b0cc939dec200efc3bf6db55b402c67ed678c7/docker directory.

Additionally, I missed some stacks like Documentation and some external services like Elasticsearch. Would be nice to include the documentation for them as well.

Thanks again for this contribution and for being conducive to my feedback.

**Comment by zealshah29** [comment]
> I search our docs and it turns out unfortunately, we do not have docs for that, which is a big miss on our end (and a potential opportunity for you to submit a PR for, once you've set it up for yourself).
> 
> Linting locally requires a bunch of tooling set up like [`just`](https://docs.openverse.org/general/general_setup.html#just), [Docker](https://docs.openverse.org/general/general_setup.html#docker) and [Node.js](https://docs.openverse.org/general/general_setup.html#node-js). Would you be open to installing them? If so, please do so and then run `just lint` from the project root directory.
> 
> If you would like to not set them up, let us know and we'll lint the file for you.

Okay, I shall do the setup and try that on my own. In case I need help, I'll let you know. 


> Yes, a separate section with a name like "External services" would be good to group these services.

Okay

> All of our external services are in the https://github.com/WordPress/openverse/tree/b4b0cc939dec200efc3bf6db55b402c67ed678c7/docker directory.
> 
> Additionally, I missed some stacks like Documentation and some external services like Elasticsearch. Would be nice to include the documentation for them as well.

Okay sure will go through it. So I'm planning to add them in the following manner:

External Services:
- cache: [Redis]()
- upstream and API database: [PostgreSQL]()
- analytics: [third-party, Plausible]()
- ...
The tools and frameworks used will be linked to their respective documentation. Would that be good?

> Thanks again for this contribution and for being conducive to my feedback.

Thank you for your guidance and support.

**Comment by zealshah29** [comment]
> 3.  Automation is not technically a part of the stack as it's more concerned with general repository management. I would drop it but I'll defer to other maintainers for their views.

Also, shall I remove Automations from stack.md?

**Comment by zealshah29** [comment]
> I search our docs and it turns out unfortunately, we do not have docs for that, which is a big miss on our end (and a potential opportunity for you to submit a PR for, once you've set it up for yourself).
> 
> Linting locally requires a bunch of tooling set up like [`just`](https://docs.openverse.org/general/general_setup.html#just), [Docker](https://docs.openverse.org/general/general_setup.html#docker) and [Node.js](https://docs.openverse.org/general/general_setup.html#node-js). Would you be open to installing them? If so, please do so and then run `just lint` from the project root directory.
> 
> If you would like to not set them up, let us know and we'll lint the file for you.

Hi so I tried installing just, but I am facing the following issue: 
image

Also, I checked the justfile and I read that it's not compatible with Windows OS. Since I am using windows OS (git bash), is there any way to do linting on my system?

**Comment by dhruvkb** [comment]
@zealshah29 Openverse development is only supported for Linux or macOS, largely because none of the core maintainers have Windows. For Windows, [we recommend WSL](https://docs.openverse.org/general/general_setup.html#operating-system) to get a Linux-like environment which is much more conducive for development.

However, in this case, we're happy to take over the PR and lint it ourselves.

**Comment by zealshah29** [comment]
Hi, so I have updated the files as per the review.

File: documentation/general/stack.md
Comment (zealshah29) [reply documentation/general/stack.md:27]: Do I need to make any changes related to the link over here?

**Comment by stacimc** [comment]
Thanks for making those updates, @zealshah29! I think this is just about ready, but I've just noticed your PR is from your `main` branch. Could you please move your changes to a new feature branch and make a new PR from it, per the [Openverse development guidelines](https://docs.openverse.org/meta/contribution/dev_flow.html)? Let us know if you have any questions. Once that's done I'd be happy to push the linting changes for you and approve!

**Review by stacimc — APPROVED** [review]
On second thought, since this PR has been open for awhile and has a good bit of context, I'll approve as is! Please use a feature branch in the future, however.

Thanks again for your contribution!

**Comment by zealshah29** [comment]
> On second thought, since this PR has been open for awhile and has a good bit of context, I'll approve as is! Please use a feature branch in the future, however.
> 
> Thanks again for your contribution!

Thank you! Sure I will take care of that from next time.

**Pull request merged**
